# *Setup*
Instalación y configuración de las herramientas necesarias.

In [2]:
!apt-get install openjdk-8-jdk
!apt install unzip
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
!pip install pyspark

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  fonts-dejavu-core fonts-dejavu-extra libatk-wrapper-java
  libatk-wrapper-java-jni libfontenc1 libgail-common libgail18 libgtk2.0-0
  libgtk2.0-bin libgtk2.0-common libxkbfile1 libxtst6 libxxf86dga1
  openjdk-8-jdk-headless openjdk-8-jre openjdk-8-jre-headless x11-utils
Suggested packages:
  gvfs openjdk-8-demo openjdk-8-source visualvm libnss-mdns
  fonts-ipafont-gothic fonts-ipafont-mincho fonts-wqy-microhei
  fonts-wqy-zenhei fonts-indic mesa-utils
The following NEW packages will be installed:
  fonts-dejavu-core fonts-dejavu-extra libatk-wrapper-java
  libatk-wrapper-java-jni libfontenc1 libgail-common libgail18 libgtk2.0-0
  libgtk2.0-bin libgtk2.0-common libxkbfile1 libxtst6 libxxf86dga1
  openjdk-8-jdk openjdk-8-jdk-headless openjdk-8-jre openjdk-8-jre-headless
  x11-utils
0 upgraded, 18 newly installed, 0 to remove and 34 not upg

In [3]:
!apt install unzip

Reading package lists... Done
Building dependency tree       
Reading state information... Done
unzip is already the newest version (6.0-25ubuntu1.1).
0 upgraded, 0 newly installed, 0 to remove and 34 not upgraded.


In [4]:
import json
import os
from pyspark import SparkContext

In [5]:
sc = SparkContext()

# **Descargamos los datasets de bicimad**
Descargaremos cada dataset de la web oficial de biciMAD. Los renombraremos a '.zip' para poder descomprimirlos y una vez descomprimidos los movemos a la carpeta 'datasets'. Por último eliminamos los ficheros comprimidos.

In [6]:
#Estructura de carpetas para el dataset
!mkdir dataset
!cd dataset/
!mkdir dataset/usages
!mkdir dataset/stations

## Datasets de uso
Aquí nos descargamos los datasets de uso por usuario y los movemos dentro de la carpeta _usages/_ dentro de _datasets/_

In [7]:
#Enero de 2019
!wget -N "https://opendata.emtmadrid.es/getattachment/2ebcc70a-4914-43c6-9ada-a0f0520032a4/201901_Usage_Bicimad.aspx"
!mv 201901_Usage_Bicimad.aspx 201901_Usage_Bicimad.zip
!unzip 201901_Usage_Bicimad.zip
!mv 201901_Usage_Bicimad.json dataset/usages
!rm 201901_Usage_Bicimad.zip

--2023-05-31 18:03:33--  https://opendata.emtmadrid.es/getattachment/2ebcc70a-4914-43c6-9ada-a0f0520032a4/201901_Usage_Bicimad.aspx
Resolving opendata.emtmadrid.es (opendata.emtmadrid.es)... 185.79.130.130
Connecting to opendata.emtmadrid.es (opendata.emtmadrid.es)|185.79.130.130|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 78311972 (75M) [application/zip]
Saving to: ‘201901_Usage_Bicimad.aspx’

201901_Usage_Bicima 100%[===================>]  74.68M  4.44MB/s    in 16s     

2023-05-31 18:03:51 (4.62 MB/s) - ‘201901_Usage_Bicimad.aspx’ saved [78311972/78311972]

Archive:  201901_Usage_Bicimad.zip
  inflating: 201901_Usage_Bicimad.json  


In [8]:
#Febrero de 2019
!wget -N "https://opendata.emtmadrid.es/getattachment/aa8c34d2-ddba-46d9-b6db-882c0b4a12f0/201902_Usage_Bicimad.aspx"
!mv 201902_Usage_Bicimad.aspx 201902_Usage_Bicimad.zip 
!unzip 201902_Usage_Bicimad.zip
!mv 201902_Usage_Bicimad.json dataset/usages
!rm 201902_Usage_Bicimad.zip

--2023-05-31 18:04:05--  https://opendata.emtmadrid.es/getattachment/aa8c34d2-ddba-46d9-b6db-882c0b4a12f0/201902_Usage_Bicimad.aspx
Resolving opendata.emtmadrid.es (opendata.emtmadrid.es)... 185.79.130.130
Connecting to opendata.emtmadrid.es (opendata.emtmadrid.es)|185.79.130.130|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 64069335 (61M) [application/zip]
Saving to: ‘201902_Usage_Bicimad.aspx’

201902_Usage_Bicima 100%[===================>]  61.10M  5.04MB/s    in 12s     

2023-05-31 18:04:18 (4.89 MB/s) - ‘201902_Usage_Bicimad.aspx’ saved [64069335/64069335]

Archive:  201902_Usage_Bicimad.zip
  inflating: 201902_Usage_Bicimad.json  


In [9]:
#Junio de 2019
!wget -N "https://opendata.emtmadrid.es/getattachment/7517a650-ccdf-4ab1-b1b0-a1d13694472e/201906_Usage_Bicimad.aspx"
!mv 201906_Usage_Bicimad.aspx 201906_Usage_Bicimad.zip
!unzip 201906_Usage_Bicimad.zip
!mv 201906_Usage_Bicimad.json dataset/usages
!rm 201906_Usage_Bicimad.zip

--2023-05-31 18:04:26--  https://opendata.emtmadrid.es/getattachment/7517a650-ccdf-4ab1-b1b0-a1d13694472e/201906_Usage_Bicimad.aspx
Resolving opendata.emtmadrid.es (opendata.emtmadrid.es)... 185.79.130.130
Connecting to opendata.emtmadrid.es (opendata.emtmadrid.es)|185.79.130.130|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 59678050 (57M) [application/zip]
Saving to: ‘201906_Usage_Bicimad.aspx’

201906_Usage_Bicima 100%[===================>]  56.91M  4.79MB/s    in 12s     

2023-05-31 18:04:39 (4.60 MB/s) - ‘201906_Usage_Bicimad.aspx’ saved [59678050/59678050]

Archive:  201906_Usage_Bicimad.zip
  inflating: 201906_Usage_Bicimad.json  


## Datasets de estaciones
Aquí nos descargamos los datasets de ocupacion por estación y los movemos dentro de la carpeta _stations/_ dentro de _datasets/_

In [10]:
#Situaciones estaciones Enero 2019
!wget -N "https://opendata.emtmadrid.es/getattachment/17206b97-1de9-4c95-a20f-3331d699d2fa/Bicimad_Stations_201901.aspx"
!mv Bicimad_Stations_201901.aspx Bicimad_Stations_201901.zip 
!unzip Bicimad_Stations_201901.zip
!mv Bicimad_Stations_201901.json dataset/stations/
!rm Bicimad_Stations_201901.zip




--2023-05-31 18:04:52--  https://opendata.emtmadrid.es/getattachment/17206b97-1de9-4c95-a20f-3331d699d2fa/Bicimad_Stations_201901.aspx
Resolving opendata.emtmadrid.es (opendata.emtmadrid.es)... 185.79.130.130
Connecting to opendata.emtmadrid.es (opendata.emtmadrid.es)|185.79.130.130|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5217347 (5.0M) [application/zip]
Saving to: ‘Bicimad_Stations_201901.aspx’

Bicimad_Stations_20 100%[===================>]   4.98M  3.38MB/s    in 1.5s    

2023-05-31 18:04:55 (3.38 MB/s) - ‘Bicimad_Stations_201901.aspx’ saved [5217347/5217347]

Archive:  Bicimad_Stations_201901.zip
  inflating: Bicimad_Stations_201901.json  


# **Creamos los RDD**
Una vez tenemos las bases de datos descargadas vamos a codificarlas de forma cómoda.

-----
Para tener los datos almacenados de una forma cómoda primero creamos un diccionario cuya clave va a ser un string con el mes y el año del dataset y como valor va a tener el rdd asociado al uso por usuario de ese mes.

La funcion ```mapper_usages``` nos servirá para crear la rdd mas legible. Nos quedaremos con los datos necesarios y cada linea la codificaremos como un diccionario.

In [11]:
def mapper_usages(line):
  data = json.loads(line)
  user = data['user_type']
  user_day = data['user_day_code']
  start = data['idunplug_station']
  end = data['idplug_station']
  date = data['unplug_hourTime']['$date'][0:10]
  hora = data['unplug_hourTime']['$date'][11:19]
  time = data['travel_time']
  age = data['ageRange']
  return user,user_day, start, end , time, date, hora, age

In [12]:
rdd_usages={}

In [13]:
directory = 'dataset/usages'
for filename in os.listdir(directory):
    if filename.endswith(".json"):
      #Nos quedamos con la fecha del dataset en formato YYYYMM
      name = filename.split("_")[0]
      rdd_usages[name] = \
      sc.textFile(os.path.join(directory, filename)).map(mapper_usages)
      #DEBUG starts
      print(name)
      #DEBUG ends
    else:
        continue

201902
201901
201906


In [14]:
rdd_usages

{'201902': PythonRDD[6] at RDD at PythonRDD.scala:53,
 '201901': PythonRDD[7] at RDD at PythonRDD.scala:53,
 '201906': PythonRDD[8] at RDD at PythonRDD.scala:53}

----
Por otro lado un diccionario cuya clave va a ser tambien un string con el mes y el año del dataset y como valor va a tener el rdd asociado a la ocupacion de las estaciones en ese mes.

Aquí usaremos la funcion ```mapper_stations``` para acomodar los datos.

In [15]:
rdd_stations = {} 

In [16]:
def mapper_stations(line):
  data = json.loads(line)
  day = data['_id'][0:10]
  hour = data['_id'][11:27]
  station = data['stations']
  return {"day": day, "hour": hour, "station": station}

In [17]:
directory = 'dataset/stations'
for filename in os.listdir(directory):
    if filename.endswith(".json"):
      #Nos quedamos con la fecha del dataset en formato YYYYMM
      name = filename.split("_")[2].split(".")[0]
      rdd_stations[name] = \
      sc.textFile(os.path.join(directory, filename)).map(mapper_stations)
      #DEBUG starts
      print(name)
      #DEBUG ends
    else:
        continue

201901


Como usar la nueva forma de organizar los rdd

In [ ]:
rdd_stations["201901"].take(1)

[{'day': '2019-01-01',
  'hour': '00:50:23.009468',
  'station': [{'activate': 1,
    'name': 'Puerta del Sol A',
    'reservations_count': 0,
    'light': 3,
    'total_bases': 24,
    'free_bases': 0,
    'number': '1a',
    'longitude': '-3.7024255',
    'no_available': 1,
    'address': 'Puerta del Sol n� 1',
    'latitude': '40.4168961',
    'dock_bikes': 0,
    'id': 1},
   {'activate': 1,
    'name': 'Puerta del Sol B',
    'reservations_count': 0,
    'light': 3,
    'total_bases': 24,
    'free_bases': 0,
    'number': '1b',
    'longitude': '-3.7024207',
    'no_available': 1,
    'address': 'Puerta del Sol n� 1',
    'latitude': '40.4170009',
    'dock_bikes': 0,
    'id': 2},
   {'activate': 1,
    'name': 'Miguel Moya',
    'reservations_count': 0,
    'light': 0,
    'total_bases': 24,
    'free_bases': 20,
    'number': '2',
    'longitude': '-3.7058415',
    'no_available': 0,
    'address': 'Calle Miguel Moya n� 1',
    'latitude': '40.4205886',
    'dock_bikes': 2,
  

In [ ]:
rdd_usages["201901"].take(5)

[(1,
  'b3b5b61509ab89afbf4d1e589d229bebe6637c2de819f5d612762c939b8d7f1f',
  82,
  83,
  162,
  '2019-01-01',
  '00:00:00',
  0),
 (1,
  'a18550d468153bb8e1808d4f0a85955579951832772996de08a154d7b4b11ee0',
  133,
  36,
  415,
  '2019-01-01',
  '00:00:00',
  5),
 (1,
  '905737396c2fe343c84d62eaee3ab66e19632638c5bea772b1e82a9d68f6a8b0',
  25,
  90,
  545,
  '2019-01-01',
  '00:00:00',
  0),
 (1,
  '5922b4a1a2591161ebf1ad66f42f52cc7796f3754c2c76c52313a01c5e4dbd96',
  105,
  155,
  417,
  '2019-01-01',
  '00:00:00',
  3),
 (1,
  'ed1c1a22c120e7d42cc522c33ab41db549791f3914b273a1f3926d62f6b9203a',
  51,
  44,
  448,
  '2019-01-01',
  '00:00:00',
  4)]

# Usuarios por tipos

##Problema1
Analizamos los tipos de usuarios. 
Queremos saber que tipo de usuarios utilizan más Bicimad. Para ello analizamos la cantidad de usuarios de cada tipo. Utilizamos varios meses de 2019, para poder comprobar si es algo casual o no.

In [ ]:
rdd_usages["201901"].countByKey()

defaultdict(int, {1: 253661, 2: 1967, 3: 21746, 0: 1849})

In [ ]:
rdd_usages["201902"].countByKey()

defaultdict(int, {3: 21734, 1: 278366, 2: 3862})

In [ ]:
rdd_usages["201906"].countByKey()

defaultdict(int, {1: 418759, 2: 5738, 3: 26314})

###Conclusiones
Podemos ver como en todos los meses la mayor parte de los usuarios son del tipo 1.
Los números nos indican: 
0: No se ha podido determinar el tipo de usuario
1: Usuario anual (poseedor de un pase anual)
2: Usuario ocasional
3: Trabajador de la empresa
Por tanto podemos concluir que la mayor parte de los usuarios de Bicimad entre Enero de 2019 y Junio de 2019 son usuarios anuales. 

##Estudio de los usuarios tipo 1
Una vez henmos concluido cuales son los usuarios habuituales de Bicimad, vamos a estudiar unicamente estos.

In [18]:
selected_type = 1
rdd_users = rdd_usages["201901"].filter(lambda x: x[0]==selected_type).map(lambda x: (x[1],tuple(x[2:])))

In [ ]:
rdd_users.count()

253661

In [ ]:
rdd_users.countByKey()

defaultdict(int,
            {'b3b5b61509ab89afbf4d1e589d229bebe6637c2de819f5d612762c939b8d7f1f': 1,
             'a18550d468153bb8e1808d4f0a85955579951832772996de08a154d7b4b11ee0': 1,
             '905737396c2fe343c84d62eaee3ab66e19632638c5bea772b1e82a9d68f6a8b0': 1,
             '5922b4a1a2591161ebf1ad66f42f52cc7796f3754c2c76c52313a01c5e4dbd96': 1,
             'ed1c1a22c120e7d42cc522c33ab41db549791f3914b273a1f3926d62f6b9203a': 1,
             'c7d4e8f50d4a318e4f9fb08d4b715dc9015be1f415358b7c1afc5567399d7fe3': 1,
             '3c9e487b04b4b25e0f885a93daa0cdf4b0072835280e52d9edf52327ea647362': 1,
             'b70eb1a71e53c6351b4835f618a459886e2045a44635e489589dc6d20a501ea6': 1,
             '1ad9b018315170d824bb396889046eaeff8ce0a0be46b3054331594ff04ce07c': 1,
             'f429ea17c7b9f0fbff82dc4ee2ea1306794b873ec4540704db0d9f99b6c0046d': 1,
             '295ac2a11ec7c646dd330331a1e344a0ba3309456031741b358b8d19fbae9356': 1,
             '1c77fcb20cd4db492d10fa16b4930882aec4ffc09f3a5

In [ ]:
rdd_users.take(1)

[('b3b5b61509ab89afbf4d1e589d229bebe6637c2de819f5d612762c939b8d7f1f',
  (82, 83, 162, '2019-01-01', '00:00:00', 0))]

## Usuarios por grupos de edades.(Problema 2)
Queremos analizar el uso de bicimad en función de la edad.
0: No se ha podido determinar el rango de edad del usuario.
1: El usuario tiene entre 0 y 16 años.
2: El usuario tiene entre 17 y 18 años.
3: El usuario tiene entre 19 y 26 años.
4: El usuario tiene entre 27 y 40 años.
5: El usuario tiene entre 41 y 65 años.
6: El usuario tiene 66 años o más.

###Edades Enero 2019

In [ ]:
selected_age = 0
rdd_ages=rdd_usages["201901"].filter(lambda x: x[7]==selected_age).map(lambda x: (x[0],tuple(x[1:7])))

In [ ]:
rdd_ages.count()

112805

In [ ]:
selected_age=3
rdd_ages.count()

15716

In [ ]:
selected_age=4
rdd_ages.count()

82092

In [ ]:
selected_age=2
rdd_ages.count()

1926

In [ ]:
selected_age=1
rdd_ages.count()

3023

In [ ]:
selected_age=5
rdd_ages.count()

62365

In [ ]:
selected_age=6
rdd_ages.count()

1296

En enero de 2019 el grupo de edad con un rango determinado que más utilizó bicimad fue el grupo 4 (entre 27 y 40 años).

###Edades resto de meses.
Ahora estudiamos la cantidad de usuario que usaron Bicimad el resto de meses y pertenecen a este grupo de edad.

####Febrero 2019

In [ ]:
selected_age = 4
rdd_ages02=rdd_usages["201902"].filter(lambda x: x[7]==selected_age).map(lambda x: (x[0],tuple(x[1:7])))
rdd_ages02.count()

87506

In [ ]:
selected_age = 5
rdd_ages02=rdd_usages["201902"].filter(lambda x: x[7]==selected_age).map(lambda x: (x[0],tuple(x[1:7])))
rdd_ages02.count()

67231

In [ ]:
selected_age = 3
rdd_ages02=rdd_usages["201902"].filter(lambda x: x[7]==selected_age).map(lambda x: (x[0],tuple(x[1:7])))
rdd_ages02.count()

16392

In [ ]:
selected_age = 2
rdd_ages02=rdd_usages["201902"].filter(lambda x: x[7]==selected_age).map(lambda x: (x[0],tuple(x[1:7])))
rdd_ages02.count()

1628

####Junio 2019

In [ ]:
selected_age = 4
rdd_ages06=rdd_usages["201906"].filter(lambda x: x[7]==selected_age).map(lambda x: (x[0],tuple(x[1:7])))
rdd_ages06.count()

124219

In [ ]:
selected_age = 5
rdd_ages06=rdd_usages["201906"].filter(lambda x: x[7]==selected_age).map(lambda x: (x[0],tuple(x[1:7])))
rdd_ages06.count()

95046

In [ ]:
selected_age = 3
rdd_ages06=rdd_usages["201906"].filter(lambda x: x[7]==selected_age).map(lambda x: (x[0],tuple(x[1:7])))
rdd_ages06.count()

23712

##Conclusión
El tipo de usuario según la edad de Bicimad tiene en mayoría entre 27 y 40 años. 

# **Nos quedamos con el origen y el destino de los viajes.** 

In [19]:
rdd_trips = rdd_users.map(lambda x: (x[1][0], x[1][1]))

In [ ]:
rdd_trips.take(1)

[(82, 83)]

##Problema3
Planteamos en problema 3 como el estudio de los viajes según estación de origen.


* Si queremos saber cuantos viajes se han hecho desde una cierta estación de origen:

In [ ]:
origen=5
rdd_trips.filter(lambda x: x[0]==origen).count()

1146

In [ ]:
origen=34
rdd_trips.filter(lambda x: x[0]==origen).count()

858

Estudiamos las estaciones más utilizadas como origen.



In [20]:
rdd_origen= rdd_users.map(lambda x: (x[1][0],1))

In [21]:
cout=rdd_origen.countByKey()
rdd_origenl= sc.parallelize(list(cout.items()))
rdd_origen_ord=rdd_origenl.sortBy(lambda x: -x[1])
rdd_origen_ord.take(3)

[(163, 3742), (175, 3461), (57, 3360)]

Estación de origen más utilizada 163.

Las menos utilizadas como origen.

In [23]:
cout=rdd_origen.countByKey()
rdd_origenl= sc.parallelize(list(cout.items()))
rdd_origen_ord2=rdd_origenl.sortBy(lambda x: x[1])
rdd_origen_ord2.take(4)

[(2008, 1), (28, 246), (29, 326), (119, 374)]

Las estación menos utilizada como origen es la 28.

* Queremos saber si hay algún viaje con cierto origen y destino.

In [ ]:
origen=5
destino=2
rdd_trips.filter(lambda x: x[0]==origen and x[1]==destino).count()

0

In [ ]:
rdd_trips.take(1)

[(82, 83)]

In [ ]:
origen=82
destino=83
rdd_trips.filter(lambda x: x[0]==origen and x[1]==destino).count()

17

#Recorrido de los usuarios

* De los usuarios que han realizado 2 viajes en un día queremos ver cuantos han cogido la bici para realizar el segundo viaje en la estación de destino del primer viaje.

In [ ]:
rdd_users.groupByKey().mapValues(tuple).filter(lambda x: len(x[1])==2).take(1)

[('34a66a0165199d0dbc90782dafe9c5089f02211701894b3015d90a1fde6a632b',
  ((9, 71, 505, '2019-01-01', '01:00:00', 4),
   (71, 164, 1141, '2019-01-01', '05:00:00', 4)))]

In [ ]:
def viajes(data):
  viaje1= data[1][0]
  viaje2=data[1][1]
  return viaje1[1]==viaje2[0]

In [ ]:
rdd_viajedoble= rdd_users.groupByKey().mapValues(tuple).filter(lambda x: len(x[1])==2).filter(viajes)
rdd_viajedoble.count()

30638

In [ ]:
rdd_viajedoble.take(2)

[('34a66a0165199d0dbc90782dafe9c5089f02211701894b3015d90a1fde6a632b',
  ((9, 71, 505, '2019-01-01', '01:00:00', 4),
   (71, 164, 1141, '2019-01-01', '05:00:00', 4))),
 ('1ca5ca46ed2b047a6e95af098b5eba1a8fa2c29d2d67a687f3e0bcab280fcfa3',
  ((123, 168, 374, '2019-01-01', '02:00:00', 4),
   (168, 124, 349, '2019-01-01', '15:00:00', 4)))]

* Ahora nos quedamos con los usuarios que han hecho 4 viajes al día.




In [ ]:
 rdd_users.groupByKey().mapValues(tuple).filter(lambda x: len(x[1])==4).take(2)

[('623cac7ccd6e01c8b52c0991f7e2b52dbcbd799ec68da49cb91e8ed0fcf01d84',
  ((107, 13, 521, '2019-01-01', '01:00:00', 5),
   (6, 107, 379, '2019-01-01', '14:00:00', 5),
   (107, 6, 450, '2019-01-01', '20:00:00', 5),
   (13, 118, 297, '2019-01-01', '23:00:00', 5))),
 ('be5aad29df9e7f7efac9e468e8c6abcb2a83279d094d655d8da3b11a2f873a99',
  ((40, 124, 283, '2019-01-02', '09:00:00', 4),
   (123, 130, 461, '2019-01-02', '14:00:00', 4),
   (130, 123, 429, '2019-01-02', '15:00:00', 4),
   (124, 10, 377, '2019-01-02', '18:00:00', 4)))]

Estudiamos cuantos de ellos tienen al menos un vaje de ida y vuelta.

In [ ]:
def idavuelta(data):
    viaje1= data[1][0]
    viaje2=data[1][1]
    viaje3=data[1][2]
    viaje4=data[1][3]
    return (viaje1[0]==viaje2[1] and viaje1[1]==viaje2[0]) or \
            (viaje2[0]==viaje3[1] and viaje2[1]==viaje3[0]) or \
              (viaje3[0]==viaje4[1] and viaje3[1]==viaje4[0])

In [ ]:
rdd_idavuelta = rdd_users.groupByKey().mapValues(tuple).filter(lambda x: len(x[1])==4).filter(idavuelta)

In [ ]:
rdd_idavuelta.count()

2671

In [ ]:
rdd_idavuelta.take(1)

[('623cac7ccd6e01c8b52c0991f7e2b52dbcbd799ec68da49cb91e8ed0fcf01d84',
  ((107, 13, 521, '2019-01-01', '01:00:00', 5),
   (6, 107, 379, '2019-01-01', '14:00:00', 5),
   (107, 6, 450, '2019-01-01', '20:00:00', 5),
   (13, 118, 297, '2019-01-01', '23:00:00', 5)))]

# Encontrar una bici en una estación a cierta hora (Problema 4) 

In [ ]:
#Hagamos una lista con los datos de todas las estaciones  
def crearDiEst(lista):
  dicEstac = {}
  for station in lista:
    dicEstac[station['number']] = {}
    dicEstac[station['number']]['id'] = station['id']
    dicEstac[station['number']]['latitude'] = station['latitude']
    dicEstac[station['number']]['longitude'] = station['longitude']
    dicEstac[station['number']]['name'] = station['name']
    dicEstac[station['number']]['total_bases'] = station['total_bases']
  return dicEstac

In [ ]:
listEstacion = (rdd_stations["201901"]).map(lambda x: crearDiEst(x['station'])).take(1)

In [ ]:
dicEstaciones = listEstacion[0]

In [ ]:
dicEstaciones['5']

{'id': 6,
 'latitude': '40.4285280',
 'longitude': '-3.7020599',
 'name': 'Fuencarral',
 'total_bases': 27}

In [ ]:
#Definimos la hora y el origen
origen = '57'
hora = '03:00:00'

In [ ]:
#De todas las estaciones que hay en todos los datos vamos a intentar quedarnos 
#tan solo con los que son de la estacion que queremos
def estaciones(lista):
  filtro = []
  # = json.loads(lista)
  for estacion in lista:
    if estacion['number'] == origen:
      filtro.append(estacion)
  return filtro

In [ ]:
rddE_Sit_fil = rdd_stations["201901"].map(lambda x: {"day" : x["day"], "hour" : x["hour"], "stations" : estaciones(x["station"])})
rddE_Sit_fil.take(3)

[{'day': '2019-01-01',
  'hour': '00:50:23.009468',
  'stations': [{'activate': 1,
    'name': 'Santa Cruz del Marcenado',
    'reservations_count': 0,
    'light': 2,
    'total_bases': 24,
    'free_bases': 10,
    'number': '57',
    'longitude': '-3.7126299',
    'no_available': 0,
    'address': 'Calle Santa Cruz del Marcenado n� 24',
    'latitude': '40.4295658',
    'dock_bikes': 12,
    'id': 61}]},
 {'day': '2019-01-01',
  'hour': '01:50:30.148211',
  'stations': [{'activate': 1,
    'name': 'Santa Cruz del Marcenado',
    'reservations_count': 0,
    'light': 2,
    'total_bases': 24,
    'free_bases': 10,
    'number': '57',
    'longitude': '-3.7126299',
    'no_available': 0,
    'address': 'Calle Santa Cruz del Marcenado n� 24',
    'latitude': '40.4295658',
    'dock_bikes': 12,
    'id': 61}]},
 {'day': '2019-01-01',
  'hour': '02:50:27.594144',
  'stations': [{'activate': 1,
    'name': 'Santa Cruz del Marcenado',
    'reservations_count': 0,
    'light': 2,
    'total

Nos fijamos que a las 2:50 había 12 bicis ancladas en la estación deseada ('dock_bikes'). Por tanto, es muy probable que haya bicis disponibles para realizar nuestro viaje.

# Anális del tiempo de los viajes 

Nos quedamos con los tiempos de cada viaje.

In [ ]:
rdd_time=rdd_users.map(lambda x: (x[1][2]))

Queremos saber los viajes que han durado más de media hora.

In [ ]:
rdd_time.filter(lambda x: x>1800).count()

13859

Viajes que han durado más de una hora.

In [ ]:
rdd_time.filter(lambda x: x>3600).count()

5038

In [ ]:
rdd_3600=rdd_users.map(lambda x: (x[1][0],x[1][1],x[1][2])).filter(lambda x: x[2]>3600)

In [ ]:
rdd_3600.take(10)

[(49, 12, 6365),
 (136, 50, 9860),
 (164, 84, 8581),
 (62, 62, 15773),
 (54, 51, 5660),
 (52, 175, 3678),
 (52, 50, 4005),
 (139, 139, 5841),
 (4, 82, 14867),
 (31, 91, 9799)]

In [ ]:
rdd_3600.count()

5038